src/dtls.c

```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+            }
+
+            /* Supported groups */
+            ret = TlsxFindByType(&tlsx, TLSX_SUPPORTED_GROUPS,
+                                 ch->extension);
+            if (ret != 0)
+                goto dtls13_cleanup;
+            if (tlsx.size != 0) {
+                ret = TLSX_SupportedCurve_Parse(ssl, tlsx.elements,
+                                                tlsx.size, 1, &parsedExts);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+            }
+
+            /* Key share */
+            ret = TlsxFindByType(&tlsx, TLSX_KEY_SHARE,
+                                 ch->extension);
+            if (ret != 0)
+                goto dtls13_cleanup;
+            if (tlsx.size != 0) {
+                ret = TLSX_KeyShare_Parse_ClientHello(ssl, tlsx.elements,
+                                          tlsx.size, &parsedExts);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+                haveKS = 1;
+#endif
+            }
+
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+            /* Pre-shared key */
+            ret = TlsxFindByType(&tlsx, TLSX_PRE_SHARED_KEY, ch->extension);
+            if (ret != 0)
+                goto dtls13_cleanup;
+            if (tlsx.size != 0) {
+                /* Let's just assume that the binders are correct here. We will
+                 * actually verify this in the stateful part of the processing
+                 * and if they don't match we will error out there anyway. */
+                byte modes;
+
+                ret = TlsxFindByType(&tlsx, TLSX_PSK_KEY_EXCHANGE_MODES,
+                                     ch->extension);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+                if (tlsx.size == 0)
+                    ERROR_OUT(MISSING_HANDSHAKE_DATA, dtls13_cleanup);
+                ret = TLSX_PskKeyModes_Parse_Modes(tlsx.elements, tlsx.size,
+                                                  client_hello, &modes);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+                if ((modes & (1 << PSK_DHE_KE)) && !ssl->options.noPskDheKe) {
+                    if (!haveKS)
+                        ERROR_OUT(MISSING_HANDSHAKE_DATA, dtls13_cleanup);
+                    doKE = 1;
+                }
+                else {
+                    if ((modes & (1 << PSK_KE)) == 0)
+                        ERROR_OUT(PSK_KEY_ERROR, dtls13_cleanup);
+                }
+            }
+#endif
+
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+            if (usePSK) {
+                if (pskInfo->isValid) {
+                    cs.cipherSuite0 = pskInfo->cipherSuite0;
+                    cs.cipherSuite = pskInfo->cipherSuite;
+                }
+                else {
+                    /* Only support the default ciphersuite for PSK */
+                    cs.cipherSuite0 = TLS13_BYTE;
```

Comment:
The PSK does not have to be a ticket. It can also be an identity provided to the user so that they can supply their own key material. Here we assume that the PSK is valid and offer the default PSK cipher suite. This mimics the logic in our stateful logic (obviously there we check if the PSK is valid).

Reply:
So even if `pskInfo->isValid` is false we still have a valid hardcoded/predefined "psk"? This is a little confusing. 
 
In that case, IIRC the ciphersuite is returned by one version of the psk callback. Should we use that value?

Reply:
Good feedback. I implemented asking the callback and falling back to a full handshake in the last commit.

Reply:
Good feedback. I implemented asking the callback and falling back to a full handshake in the last commit.